### PR TITLE
connector/ws: allow reading token from file in addition to config

### PR DESCRIPTION
**Pull request by lschuermann**
This makes it easier to exclude secrets from otherwise public configuration files.

### CODE REVIEW

File: connector/ws/src/lib.rs [hunk 1]
```diff
@@ -82,8 +90,30 @@ impl<S: connector::Supervisor> Inner<S> {
...
+                    WsConnectorError::Config(ConfigError::IoError(token_file.clone(), io_err))
+                })?;
+
+                let token = base64::prelude::BASE64_STANDARD
+                    .decode(&token_base64)
+                    .map_err(|_| ())
+                    .and_then(|token_bytes| Ok(AuthToken(token_bytes.try_into().map_err(|_| ())?)))
+                    .map_err(|_| {
+                        WsConnectorError::Config(ConfigError::InvalidToken(
+                            "Provided token is not a valid base64-encoded string of the required length"
```
Comment (max-cura) [inline connector/ws/src/lib.rs:105]: This should probably just be the `#[error()]` string.

File: connector/ws/src/lib.rs [hunk 2]
```diff
@@ -82,8 +90,30 @@ impl<S: connector::Supervisor> Inner<S> {
             .install_default()
             .unwrap();
 
-        let token_ser_string =
-            serde_json::to_string(&self.config.token).expect("failed to re-serialize token");
+        let token = match self.config.token {
+            WsConnectorConfigToken::TokenFile { ref token_file } => {
```
Comment (max-cura) [inline connector/ws/src/lib.rs:94]: Minor: would it make more sense to match on `&self.config.token` and do away with the `ref`s?

**Review by max-cura — COMMENTED** [review]
Other than comments, looks good. Would love to get this into mainline ASAP, since I'm also going through connector/ws to remove panics and audit any restart-related functionality, since surviving supervisor restart is a strict subset of the functionality required to weather switchboard restart.

File: connector/ws/src/lib.rs (hunk 1 above)
Comment (lschuermann) [reply connector/ws/src/lib.rs:105]: Okay, what would I put in this string then?

File: connector/ws/src/lib.rs (hunk 2 above)
Comment (lschuermann) [reply connector/ws/src/lib.rs:94]: Either way works for me.

File: connector/ws/src/lib.rs (hunk 1 above)
Comment (lschuermann) [reply connector/ws/src/lib.rs:105]: I mean, I can convert both of these error variants to strings and pass that through?

File: connector/ws/src/lib.rs (hunk 2 above)
Comment (max-cura) [reply connector/ws/src/lib.rs:94]: Eh, no need to change it—mostly just curious since I so rarely see `ref`.

File: connector/ws/src/lib.rs (hunk 1 above)
Comment (max-cura) [reply connector/ws/src/lib.rs:105]: Was thinking of just making `InvalidToken` a unit variant, e.g.:
```rs
enum ConfigError {
  // [snip]
  #[error("Provided token is not a valid base64-encoded string of the required length")]
  InvalidToken,
}
```

**Comment by lschuermann** [comment]
I think this needs to also strip the string of all trailing and leading whitespace, such as newlines. Let me add this!

**Comment by max-cura** [comment]
LGTM.

**Comment by max-cura** [comment]
Belatedly, I'm realizing that this might have been a use case for figment and environment variables?

**Pull request merged**